src/query/service/src/interpreters/interpreter_copy.rs

```diff
@@ -483,3 +481,190 @@ fn fill_const_columns(
+    debug!("required source schema:{:?}", plan_required_source_schema);
+    debug!("required values schema:{:?}", plan_required_values_schema);
+
+    if source_schema != plan_required_source_schema {
+        // only parquet need cast
+        let func_ctx = ctx.get_function_context()?;
+        main_pipeline.add_transform(|transform_input_port, transform_output_port| {
+            TransformCastSchema::try_create(
+                transform_input_port,
+                transform_output_port,
+                source_schema.clone(),
+                plan_required_source_schema.clone(),
+                func_ctx.clone(),
+            )
+        })?;
+    }
+
+    if !plan_values_consts.is_empty() {
+        fill_const_columns(
+            ctx.clone(),
+            main_pipeline,
+            source_schema,
+            plan_required_values_schema.clone(),
+            plan_values_consts,
+        )?;
+    }
+
+    let stage_info_clone = plan_stage_table_info.stage_info;
+    let write_mode = plan_write_mode;
+    let mut purge = true;
+    match write_mode {
+        CopyIntoTableMode::Insert { overwrite } => {
+            if use_commit {
+                append2table(
+                    ctx.clone(),
+                    to_table,
+                    plan_required_values_schema,
+                    main_pipeline,
+                    None,
+                    overwrite,
+                    AppendMode::Copy,
+                )?;
+            } else {
+                append2table_without_commit(
+                    ctx.clone(),
+                    to_table,
+                    plan_required_values_schema,
+                    main_pipeline,
+                    AppendMode::Copy,
+                )?
+            }
+        }
+        CopyIntoTableMode::Copy => {
+            if !stage_info_clone.copy_options.purge {
+                purge = false;
+            }
+
+            if use_commit {
+                let copied_files = CopyInterpreter::upsert_copied_files_request(
+                    ctx.clone(),
+                    to_table.clone(),
+                    stage_info_clone.clone(),
+                    files.clone(),
+                    plan_force,
+                )?;
+                append2table(
+                    ctx.clone(),
+                    to_table,
+                    plan_required_values_schema,
+                    main_pipeline,
+                    copied_files,
+                    false,
+                    AppendMode::Copy,
+                )?;
+            } else {
+                append2table_without_commit(
+                    ctx.clone(),
+                    to_table,
+                    plan_required_values_schema,
+                    main_pipeline,
+                    AppendMode::Copy,
+                )?
+            }
+        }
+        CopyIntoTableMode::Replace => {}
+    }
+    if local_id == ctx.get_cluster().local_id {
+        main_pipeline.set_on_finished(move |may_error| {
```

Comment:
I might have missed something here, it seems that for 
`PlanParam::CopyIntoTablePlanOption(plan)`
local_id will be `String::from("dummy")`,  thus the "on_finished" hook will NOT be set?